Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/storage/TestCertificateStorage.kt

```diff
@@ -0,0 +1,102 @@
+import androidx.core.content.edit
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import de.rki.coronawarnapp.coronatest.server.CoronaTestResult
+import de.rki.coronawarnapp.coronatest.type.CoronaTest
+import de.rki.coronawarnapp.coronatest.type.TestCertificateContainer
+import de.rki.coronawarnapp.coronatest.type.pcr.PCRCertificateContainer
+import de.rki.coronawarnapp.coronatest.type.rapidantigen.RACertificateContainer
+import de.rki.coronawarnapp.util.di.AppContext
+import de.rki.coronawarnapp.util.serialization.BaseGson
+import de.rki.coronawarnapp.vaccination.core.repository.storage.ContainerPostProcessor
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class TestCertificateStorage @Inject constructor(
+    @AppContext val context: Context,
+    @BaseGson val baseGson: Gson,
+    private val containerPostProcessor: ContainerPostProcessor,
+) {
+
+    private val prefs by lazy {
+        context.getSharedPreferences("coronatest_certificate_localdata", Context.MODE_PRIVATE)
+    }
+
+    private val gson by lazy {
+        baseGson.newBuilder().apply {
+            registerTypeAdapter(CoronaTestResult::class.java, CoronaTestResult.GsonAdapter())
+            registerTypeAdapterFactory(containerPostProcessor)
+        }.create()
+    }
+
+    private val typeTokenPCR by lazy {
+        object : TypeToken<Set<PCRCertificateContainer>>() {}.type
+    }
+
+    private val typeTokenRA by lazy {
+        object : TypeToken<Set<RACertificateContainer>>() {}.type
+    }
+
+    var testCertificates: Collection<TestCertificateContainer>
+        get() {
+            Timber.tag(TAG).d("load()")
+
+            val pcrCerts: Set<PCRCertificateContainer> = run {
+                val raw = prefs.getString(PKEY_DATA_PCR, null) ?: return@run emptySet()
+                gson.fromJson<Set<PCRCertificateContainer>>(raw, typeTokenPCR).onEach {
+                    Timber.tag(TAG).v("PCR loaded: %s", it)
+                    requireNotNull(it.identifier)
+                    requireNotNull(it.type) { "PCR type should not be null, GSON footgun." }
+                }
+            }
+
+            val raCerts: Set<RACertificateContainer> = run {
+                val raw = prefs.getString(PKEY_DATA_RA, null) ?: return@run emptySet()
+                gson.fromJson<Set<RACertificateContainer>>(raw, typeTokenRA).onEach {
+                    Timber.tag(TAG).v("RA loaded: %s", it)
+                    requireNotNull(it.identifier)
+                    requireNotNull(it.type) { "RA type should not be null, GSON footgun." }
+                }
+            }
+
+            return (pcrCerts + raCerts).also {
+                Timber.tag(TAG).v("Loaded %d certificates.", it.size)
+            }
+        }
+        set(value) {
+            Timber.tag(TAG).d("save(testCertificates=%s)", value)
+            prefs.edit {
+                value.filter { it.type == CoronaTest.Type.PCR }.run {
```

Comment:
why do you need to check for the type here ? is not the custom Gson adapter should be doing that ?